cdap-ui/app/cdap/services/datasource/index.js

```diff
@@ -0,0 +1,251 @@
+            delete this.bindings[hash];
+          }
+        }
+      }
+    );
+  }
+
+  request(resource = {}) {
+    let generatedResource = {
+      id: uuid.v4(),
+      json: resource.json || true,
+      method: resource.method || 'GET',
+      suppressErrors: resource.suppressErrors || false
+    };
+
+    if (resource.body) {
+      generatedResource.body = resource.body;
+    }
+    if (resource.data) {
+      generatedResource.body = resource.data;
+    }
+    if (resource.headers) {
+      generatedResource.headers = resource.headers;
+    }
+    if (resource.contentType) {
+      generatedResource.headers['Content-Type'] = resource.contentType;
+    }
+    if (!resource.url) {
+      resource.url = this.constructUrl(resource);
+    }
+
+    generatedResource.url = this.buildUrl(resource.url, resource.params);
+
+    let subject = new Rx.Subject();
+
+    this.bindings[generatedResource.id] = {
+      rx: subject,
+      resource: generatedResource,
+      type: 'REQUEST'
+    };
+
+    Socket.send({
+      action: 'request',
+      resource: generatedResource
+    });
+
+    return subject;
+  }
+
+  poll(resource = {}) {
+    let generatedResource = {
+      id: uuid.v4(),
+      interval: resource.interval || 10000,
+      json: resource.json || true,
+      method: resource.method || 'GET',
+      suppressErrors: resource.suppressErrors || false,
+    };
+
+    if (resource.body) {
+      generatedResource.body = resource.body;
+    }
+    if (resource.data) {
+      generatedResource.body = resource.data;
+    }
+    if (resource.headers) {
+      generatedResource.headers = resource.headers;
+    }
+
+    if (!resource.url) {
+      resource.url = this.constructUrl(resource);
+    }
+
+    generatedResource.url = this.buildUrl(resource.url, resource.params);
+
+    let subject = new Rx.Subject();
+
+    let observable = Rx.Observable.create((obs) => {
+      subject.subscribe(
+        (data) => {
+          obs.onNext(data);
+        },
+        (err) => {
+          obs.onError(err);
+        }
+      );
+
+      return () => {
+        this.stopPoll(generatedResource.id);
+        subject.dispose();
+      };
+    });
+
+    this.bindings[generatedResource.id] = {
+      rx: subject,
+      resource: generatedResource,
+      type: 'POLL'
+    };
+
+    Socket.send({
+      action: 'poll-start',
+      resource: generatedResource
+    });
+
+    return observable;
+  }
+
+  stopPoll(resourceId) {
+    let id;
+
+    if (typeof resourceId === 'object' && resourceId !== null) {
+      id = resourceId.params.pollId;
+    } else {
+      id = resourceId;
+    }
+
+    if (this.bindings[id]) {
+      Socket.send({
+        action: 'poll-stop',
+        resource: this.bindings[id].resource
+      });
+
+      this.bindings[id].rx.dispose();
+      delete this.bindings[id];
+    }
+  }
+
+  destroy() {
+    this.socketSubscription.dispose();
+
+    // stopping existing polls
+    for (let value in this.bindings) {
```

Comment:
Just noticed it we use ```for( in)``` syntax, ```for(var i=0; i<10; i++)``` syntax and ```.forEach``` syntax for iterating over loops in the same file. 

It would be better if we always use the ```.map``` and ```.reduce``` and ```.filter``` & ```forEach``` apis. They are native and would be easier to read.

In this case something like this,
```
this.bindings
  .map(value => value.resouce)
  .forEach(resource => this.stopPoll(resource)
```